contrib/initramfs/hooks/zfsunlock.in

```diff
@@ -0,0 +1,65 @@
+#!/bin/sh
+
+PREREQ="dropbear"
+
+prereqs() {
+    echo "$PREREQ"
+}
+
+case "$1" in
+    prereqs)
+        prereqs
+        exit 0
+    ;;
+esac
+
+. /usr/share/initramfs-tools/hook-functions
+
+if [ -r "${DESTDIR}/sbin/dropbear" ]; then
+	cat >"$DESTDIR"/bin/zfsunlock <<'EOF'
+#!/bin/sh
+while [ ! -e /run/zfs_fs_name ] && [ ! -e /run/zfs_unlock_complete ]; do
+	if [ ! $entry_msg_shown ]; then
+		echo -n "Waiting for zfs init script to complete, press Ctrl-C to exit into shell ..."
+		entry_msg_shown=true
+	fi
+	sleep 1;
+	echo -n "."
+done
+if [ -e /run/zfs_unlock_complete ]; then
+	exit 0
+fi
+echo
+echo "Unlocking encrypted ZFS pools"
+echo "Enter correct password or press Ctrl-C to exit into shell"
+echo
+zfs_fs_name=""
+fs_password_accepted=false
+while true; do
+	while [ $fs_password_accepted == "true" ] && [ $(cat /run/zfs_fs_name) == "$zfs_fs_name" ] && [ ! -e /run/zfs_unlock_complete ]; do
+		sleep 1;
+        done
+	if [ -e /run/zfs_unlock_complete ]; then
+		break
+	fi
+	zfs_fs_name=$(cat /run/zfs_fs_name)
+	zfs_console_askpwd_cmd=$(cat /run/zfs_console_askpwd_cmd)
+	systemd-ask-password "Encrypted ZFS password for ${zfs_fs_name}:" | \
+		/sbin/zfs load-key $zfs_fs_name
+	if [ $(/sbin/zfs get -H -ovalue keystatus $zfs_fs_name 2> /dev/null) == "available" ]; then
+		fs_password_accepted=true
+		echo "password for $zfs_fs_name accepted"
+		zfs_console_askpwd_pid=$(ps a -o pid= -o args | grep -v grep | grep $zfs_console_askpwd_cmd | cut -d ' ' -f3 | sort -n | head -n1)
+                echo -e "resuming boot sequence\nplease reconnect in a while ..."
+		if [ ! -z $zfs_console_askpwd_pid ]; then
+			kill $zfs_console_askpwd_pid
+		fi
+	else
+		echo "Wrong password, try again"
+		fs_password_accepted=false
+	fi
+done
+EOF
+	chmod 0755 "$DESTDIR"/bin/zfsunlock
+	echo "zfsunlock" >> "$DESTDIR"/etc/profile
```

Comment:
This (appending to `/etc/profile`) still feels wrong and I'd like to know if there are better options. I realize it's in the initrd, which makes it far less bad, of course, than if that were on the running system.

Reply:
I think that any mechanism that "captures" all shells on the system is a problem. Dropbear, might not be used to do encryption unlocks but to do other things in the initramfs (e.g. remotely debugging a boot process). I think sticking to what the normal "cryptroot" does should be enough: Not forcing its way into the users foreground and if a user wants to have a unlock command directly triggered with the ssh-session they can just use ssh force-commands with their ssh-key in dropbear's authorized keys: e.g. prefix the key with `command="zfsunlock" ssh-rsa ....`.

Reply:
Right, a forced command is a good way for the admin to make it automatic. Otherwise they can just run a command on login.